### PR TITLE
feat(comp: core): add component portal

### LINKED ISSUE #113: [cdk: overlay] add cdk overlay
# [cdk: overlay] add cdk overlay

- [x] I have searched the [issues](https://github.com/IduxFE/components/issues) of this repository and believe that this is not a duplicate.

对外暴露一个函数，允许调用方调用函数来创建一个浮层，且允许调用方还合适的范围内对浮层进行自定义。

## 特性

* 通过函数可以创建浮层并返回该浮层对应的实例；

* 调用方可以通过函数返回的浮层实例，对视图进行绑定；

* 调用方可以通过返回的实例，控制浮层的初始化、显示、隐藏和销毁；

* 浮层的滚动策略有四种：
  * `close` ：滚动时自动关闭浮层；
  * `reposition` ：默认行为，滚动时重新定位。

* 浮层定位依赖于 `@popperjs/core` 

* cdk会对外抛出以下事件，需要调用方在使用时将事件绑定到视图上，使得cdk可以对浮层行为进行控制：
  
  * `OverlayTriggerEvents`
    * `onClick`
    * `onMouseEnter`
    * `onMouseLeave`
    * `onFocus`
    * `onBlur`
  * `OverlayPopperEvents`
    * `onMouseEnter`
    * `onMouseLeave`
  
  > 注意：Trigger返回的事件与调用方传入 `trigger` 的值相关。

### `overlay`
```typescript
import type { Options, Placement } from '@popperjs/core'
import type { ComponentPublicInstance, ComputedRef, Ref } from 'vue'

declare type Nullable<T> = T | null

type OverlayScrollStrategy = 'close' | 'reposition'
type OverlayTrigger = 'click' | 'hover' | 'focus'
type RefElement = Nullable<HTMLElement>
type VueElement = Nullable<ComponentPublicInstance | HTMLElement>

interface OverlayTriggerEvents {
  onClick?: (event: Event) => void
  onMouseEnter?: (event: Event) => void
  onMouseLeave?: (event: Event) => void
  onFocus?: (event: Event) => void
  onBlur?: (event: Event) => void
}

interface OverlayPopperEvents {
  onMouseEnter: () => void
  onMouseLeave: () => void
}

interface OverlayOptions {
  /* The class name of the overlay container. */
  className?: string
  /**
   * Control the visibility of the overlay
   */
  visible?: boolean
  /* Scroll strategy for overlay */
  scrollStrategy: OverlayScrollStrategy
  /* Disable the overlay */
  disable?: boolean
  /**
   * The distance between the arrow and the starting point at both ends.
   *Acting when the overlay uses border-radius.
   */
  arrowOffset?: number
  /* Whether to show arrow. */
  showArrow?: boolean
  /* Alignment of floating layer. */
  placement: Placement
  /**
   * The options of popper.
   * Used when ConnectOverlayOptions cannot meet the demand.
   * Priority is higher than ConnectOverlayOptions.
   */
  popperOptions?: Partial<Options>
  /* Trigger method. */
  trigger: OverlayTrigger
  /* Whether to allow the mouse to enter the overlay. */
  allowEnter?: boolean
  /**
   * Overlay offset.
   * [Horizontal axis offset, vertical axis offset]
   */
  offset: [number, number]
  /**
   * The delay of hiding overlay.
   * Send false if you don't need it.
   */
  hideDelay: number | false
  /**
   * The delay of showing overlay.
   * Send false if you don't need it.
   */
  showDelay: number | false
}

interface OverlayInstance {
  /**
   * Initialize the overlay.
   * The life cycle of the overlay will enter mounted.
   */
  initialize: () => void
  /**
   * Show the overlay.
   * The style of the overlay container will be set to block.
   */
  show: () => void
  /**
   * Hide the overlay.
   * The style of the overlay container will be set to none.
   */
  hide: () => void
  /**
   * Destroy the overlay.
   * The life cycle of the overlay will enter beforeDestroy.
   * To show the overlay again, please recreate.
   */
  destroy: () => void
  /**
   * TODO
   * The unique id of the overlay.
   * Provide subsequent components with markings for the specified overlay treatment.
   */
  overlayId: string
  /**
   * The display status of the current overlay.
   * Control by visible and disable.
   */
  visibility: ComputedRef<boolean>
  /**
   * The truth DOM node of the overlay.
   * The caller needs to bind the variable to the view.
   */
  overlayRef: Ref<RefElement>
  /**
   * Update overlay.
   * If the overlay has not been initialized, the overlay will be initialized first, otherwise the overlay will be update directly.
   * @param options
   */
  update: (options: Partial<OverlayOptions>) => void
 /**
   * The truth DOM node of the arrow.
   * If showArrow is false, we won't return arrowRef.
   * The caller needs to bind the variable to the view.
   */
  arrowRef?: Ref<RefElement>
  /**
   * The truth DOM node of the trigger.
   * The caller needs to bind the variable to the view.
   */
  triggerRef: Ref<VueElement>
  /**
   * Manually bind to the event on the trigger.
   */
  triggerEvents: OverlayTriggerEvents
  /**
   * Manually bind to events on the overlay.
   */
  overlayEvents: OverlayPopperEvents
}
```

#### Props

| 名称      | 说明     | 类型                    | 备注                                                         |
| --------- | -------- | ----------------------- | ------------------------------------------------------------ |
| `disable` | 禁止传送 | `boolean`               |                                                              |
| `target`  | 传送位置 | `HTMLElement \| string ` | 传入类型是字符串时，会首先判断body下是否存在以该字符串为class name的元素，如果存在会直接使用该元素，否则会创建对应的元素作为容器。 |

#### Slots

| 名称      | 说明             | 备注 |
| --------- | ---------------- | ---- |
| `default` | 需要被传送的内容 | -    |

## DEMO

```vue
<template>
  <ix-button ref="triggerRef" @click="triggerEvents.onClick">Click</ix-button>
  <teleport to="body">
    <div ref="overlayRef" @mouseenter="overlayEvents.onMouseEnter()" @mouseleave="overlayEvents.onMouseLeave()">
      tooltip
    </div>
  </teleport>
</template>

<script lang="ts">
import { defineComponent, onMounted } from 'vue'
import { useOverlay } from '@idux/cdk/overlay'

export default defineComponent({
  name: 'Basic',
  setup() {
    const { initialize, overlayRef, triggerRef, triggerEvents, overlayEvents } = useOverlay({
      visible: false,
      trigger: 'click',
      placement: 'bottom',
      offset: [0, 0],
      showDelay: false,
      hideDelay: 1000,
      allowEnter: true,
    })
    onMounted(initialize)

    return { overlayRef, triggerRef, triggerEvents, overlayEvents }
  },
})
</script>
```

### PR DESCRIPTION
**Pull request by LaamGinghong**
fix #113

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows [our guidelines](https://github.com/IduxFE/components/blob/main/docs/contributing.zh.md#commit)
- [x] Tests for the changes have been added/updated or not needed
- [x] Docs and demo have been added/updated or not needed

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Component style update
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Application (the showcase website) / infrastructure changes
- [ ] Other... Please describe:

### CODE REVIEW

File: packages/components/core/portal/src/useContainer.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    container.classList.add(target)
+    document.body.appendChild(container)
+    containerHashmap[target] = container
+    return container
+  }
+
+  return useContainer
+}
+
+export default initializeContainer()
```
Comment (danranVm) [inline packages/components/core/portal/src/useContainer.ts:27]:
```suggestion
export const useContainer = initializeContainer()
```

Comment (LaamGinghong) [reply packages/components/core/portal/src/useContainer.ts:27]: 就是用default导出啊，整个文件就一个导出，非常合理

**Comment by LaamGinghong** [comment]
是不是确定把这个组件挪到cdk？

File: packages/components/core/portal/src/types.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+   * Whether disable transporting to target place.
+   */
+  disabled?: boolean
+  /**
+   * The place that element will be send.
+   * If target is a string, component will first determine whether the element exists in the body.
+   * If it exists, component will transport elements to it.
+   * If it doesn't exist, component will create it and save it.
+   */
+  target: string | HTMLElement
```
Comment (danranVm) [inline packages/components/core/portal/src/types.ts:14]:
```suggestion
  target: string | () => HTMLElement
```

**Review by danranVm — CHANGES_REQUESTED** [review]
丢到 cdk 里面吧。

File: packages/components/core/portal/src/types.ts (hunk 2 above)
Comment (LaamGinghong) [reply packages/components/core/portal/src/types.ts:14]: 你搞我吧，传个函数进来干嘛

**Comment by LaamGinghong** [comment]
重新review

File: packages/cdk/portal/index.ts [hunk 3]
```diff
@@ -0,0 +1,2 @@
+export { default } from './src/Portal.vue'
```
Comment (danranVm) [inline packages/cdk/portal/index.ts:1]:
```suggestion
export IxPortal from './src/Portal.vue'
```

**Review by danranVm — CHANGES_REQUESTED** [review]
移到 cdk 的话，要补个简单的文档和 demo ，不然 start 会报错。

**Comment by LaamGinghong** [comment]
顺便改了一下之前的文档

**Review by danranVm — APPROVED** [review]
LGTM.

**Review by danranVm — APPROVED** [review]
LGTM.

**Pull request merged**